protocol/clproto_cpp/include/clproto.h

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <stdexcept>
+#include <string>
+
+namespace clproto {
+
+class DecodingException : public std::runtime_error {
+public:
+  explicit DecodingException(const std::string& msg);
+};
+
+class NotImplementedException : public std::runtime_error {
+public:
+  explicit NotImplementedException(const std::string& msg);
+};
+
+enum MessageType {
+  UNKNOWN_MESSAGE = 0,
+  STATE_MESSAGE = 1,
+  SPATIAL_STATE_MESSAGE = 2,
+  CARTESIAN_STATE_MESSAGE = 3,
+  CARTESIAN_POSE_MESSAGE = 4,
+  CARTESIAN_TWIST_MESSAGE = 5,
+  CARTESIAN_WRENCH_MESSAGE = 6,
+  JACOBIAN_MESSAGE = 7,
+  JOINT_STATE_MESSAGE = 8,
+  JOINT_POSITIONS_MESSAGE = 9,
+  JOINT_VELOCITIES_MESSAGE = 10,
+  JOINT_TORQUES_MESSAGE = 11,
+  SHAPE_MESSAGE = 12,
+  ELLIPSOID_MESSAGE = 13,
+};
```

Comment:
I see you didn't put `Parameters` yet. Is it harder because those are templates?

Reply:
Well, despite the large PR, I didn't want to do too much at once. I am not sure yet if encoding parameter messages makes sense, or if indeed this protocol takes care of parameters for us (using string encoded messages as parameters)

Reply:
It will depend on the API we want to provide the user at the end. If we remove `Parameters` then it means the end-user has to encode decode states from string and handle the `ros2` params calls. The other way is that we keep `Parameters` but I don't think it is actually needed to encode them, just encoding their content should suffice. In `modulo` we can simply encode / decode the contents on `get` and `set` calls.